x-pack/plugins/observability/public/data_access_service.ts

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { CoreSetup, Plugin, IContextContainer } from '../../../../src/core/public';
+import { Setup, DataAccessHandlerProvider } from './typings/data_access_service';
+import { ChartDataFetcher } from '../typings/chart';
+
+export class ObservabilityDataAccessService implements Plugin<Setup, void> {
+  private handlers = new Map<string, ChartDataFetcher>();
+  private contextContainer?: IContextContainer<DataAccessHandlerProvider>;
+
+  public setup(core: CoreSetup): Setup {
+    this.contextContainer = core.context.createContextContainer();
+
+    const api: Setup = {
+      registerProvider: (pluginOpaqueId, dataType, handler) => {
+        if (!this.contextContainer) {
+          throw new Error("Context container wasn't defined");
+        }
+        this.handlers.set(dataType, this.contextContainer.createHandler(pluginOpaqueId, handler));
+      },
+      registerContext: this.contextContainer!.registerContext,
+    };
+
+    return api;
+  }
+
+  public getHandlers() {
+    return this.handlers;
+  }
```

Comment:
Do we need to expose the raw `Map` to the outside? How about we only expose?

```suggestion
  public getDataProvider(dataType: string) {
    return this.handlers.get(dataType);
  }
```

That way nothing can modify the map without going through the `registerProvider` function and the service's API becomes more self-explanatory (and auto-completion is more helpful).

Does that make sense?